File: ruby-resend-examples/examples/double_optin/webhook.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

##
# Double Opt-In: Webhook Handler
#
# Handles the email.clicked event to confirm subscriptions.
# When a user clicks the confirmation link, this webhook:
# 1. Verifies the webhook signature
# 2. Finds the contact by email
# 3. Updates the contact to unsubscribed: false
#
# This file demonstrates the webhook logic. In production,
# integrate this into your Sinatra/Rails app's webhook endpoint.
#
# @see https://resend.com/docs/dashboard/webhooks/introduction

require "bundler/setup"
require "resend"
require "dotenv/load"
require "json"

Resend.api_key = ENV.fetch("RESEND_API_KEY")

##
# Process a double opt-in webhook event
#
# @param event [Hash] The verified webhook event
# @return [Hash] Result of processing
def process_double_optin_webhook(event)
  # Only process email.clicked events
  unless event["type"] == "email.clicked"
    return { received: true, type: event["type"], message: "Event type ignored" }
  end

  audience_id = ENV.fetch("RESEND_AUDIENCE_ID")
  recipient_email = event.dig("data", "to")&.first

  unless recipient_email
    raise "No recipient email in webhook data"
  end

  puts "Confirmation click received for: #{recipient_email}"

  # Find the contact by email
  contacts = Resend::Contacts.list(audience_id)
  contact = contacts["data"]&.find { |c| c["email"] == recipient_email }

  unless contact
    raise "Contact not found: #{recipient_email}"
  end

  # Update contact to confirmed (unsubscribed: false)
  Resend::Contacts.update({
    audience_id: audience_id,
    id: contact["id"],
    unsubscribed: false
  })

  puts "Contact confirmed: #{recipient_email} (#{contact["id"]})"

  {
    received: true,
    type: event["type"],
    confirmed: true,
    email: recipient_email,
    contact_id: contact["id"]
  }
end

# Example usage (for testing with a mock event)
if __FILE__ == $PROGRAM_NAME
  puts "=== Double Opt-In: Webhook Handler ===\n\n"
  puts "This module provides the webhook processing logic."
  puts "In production, integrate this into your web application.\n\n"

  puts "Example Sinatra integration:"
  puts <<~CODE

    post '/double-optin/webhook' do
      payload = request.body.read

      # Verify webhook signature
      event = Resend::Webhooks.verify(
        payload: payload,
        headers: {
          "svix-id" => request.env["HTTP_SVIX_ID"],
          "svix-timestamp" => request.env["HTTP_SVIX_TIMESTAMP"],
          "svix-signature" => request.env["HTTP_SVIX_SIGNATURE"]
        },
        secret: ENV["RESEND_WEBHOOK_SECRET"]
      )

      # Process the event
      result = process_double_optin_webhook(event)
      json result
    end

  CODE
end

File: php-resend-examples/src/double-optin/webhook.php
<?php

/**
 * Double Opt-In: Webhook Handler
 *
 * Handles the email.clicked event to confirm subscriptions.
 * When a user clicks the confirmation link, this webhook:
 * 1. Verifies the webhook signature
 * 2. Finds the contact by email
 * 3. Updates the contact to unsubscribed: false
 *
 * Deploy this to a public URL and register it in the Resend dashboard.
 *
 * @see https://resend.com/docs/dashboard/webhooks/introduction
 */

require_once __DIR__ . '/../../vendor/autoload.php';

use Resend\Resend;

$dotenv = Dotenv\Dotenv::createImmutable(__DIR__ . '/../..');
$dotenv->load();

$resend = Resend::client($_ENV['RESEND_API_KEY']);

// Only accept POST requests
if ($_SERVER['REQUEST_METHOD'] !== 'POST') {
    http_response_code(405);
    echo json_encode(['error' => 'Method not allowed']);
    exit;
}

header('Content-Type: application/json');

$payload = file_get_contents('php://input');

// Get Svix headers
$svixId = $_SERVER['HTTP_SVIX_ID'] ?? null;
$svixTimestamp = $_SERVER['HTTP_SVIX_TIMESTAMP'] ?? null;
$svixSignature = $_SERVER['HTTP_SVIX_SIGNATURE'] ?? null;

if (!$svixId || !$svixTimestamp || !$svixSignature) {
    http_response_code(400);
    echo json_encode(['error' => 'Missing webhook headers']);
    exit;
}

$webhookSecret = $_ENV['RESEND_WEBHOOK_SECRET'] ?? null;
if (!$webhookSecret) {
    http_response_code(500);
    echo json_encode(['error' => 'Webhook secret not configured']);
    exit;
}

try {
    // Verify webhook signature
    $event = $resend->webhooks->verify([
        'payload' => $payload,
        'headers' => [
            'svix-id' => $svixId,
            'svix-timestamp' => $svixTimestamp,
            'svix-signature' => $svixSignature,
        ],
        'secret' => $webhookSecret,
    ]);

    // Only process email.clicked events
    if ($event['type'] !== 'email.clicked') {
        echo json_encode([
            'received' => true,
            'type' => $event['type'],
            'message' => 'Event ignored',
        ]);
        exit;
    }

    $audienceId = $_ENV['RESEND_AUDIENCE_ID'] ?? null;
    if (!$audienceId) {
        http_response_code(500);
        echo json_encode(['error' => 'RESEND_AUDIENCE_ID not configured']);
        exit;
    }

    // Get recipient email from webhook data
    $recipientEmail = $event['data']['to'][0] ?? null;
    if (!$recipientEmail) {
        http_response_code(400);
        echo json_encode(['error' => 'No recipient email in webhook data']);
        exit;
    }

    error_log("Confirmation click received for: {$recipientEmail}");

    // Find contact by email
    $contacts = $resend->contacts->list($audienceId);
    $contact = null;
    foreach ($contacts->data as $c) {
        if ($c->email === $recipientEmail) {
            $contact = $c;
            break;
        }
    }

    if (!$contact) {
        http_response_code(404);
        echo json_encode(['error' => 'Contact not found']);
        exit;
    }

    // Update contact to confirmed (unsubscribed: false)
    $resend->contacts->update($audienceId, $contact->id, [
        'unsubscribed' => false,
    ]);

    error_log("Contact confirmed: {$recipientEmail} ({$contact->id})");

    echo json_encode([
        'received' => true,
        'type' => $event['type'],
        'confirmed' => true,
        'email' => $recipientEmail,
        'contact_id' => $contact->id,
    ]);

} catch (Exception $e) {
    http_response_code(400);
    echo json_encode(['error' => $e->getMessage()]);
}
